### DISCUSSION TITLE
Import all .py files in a folder

### DISCUSSION THREAD

**Discussion by jhugery-blade** [micropython/micropython#15741] [Using MicroPython]
Question....  From https://docs.python.org/3/tutorial/modules.html I know that I can create an __init__.py in a folder path /foldera/folderb  and define a variable __all__ within which I define all modules that can be imported when executing:
    from foldera.folderb  import *
This does not appear to work in micropython 1.23.0 on a pico_w.  is there another way (without explicitly importing each py file) to load all .py files in a folder?

**Comment by sosi-deadeye** [comment]
**run_test_module.py**:
```python
from test_module import *

snapshot = globals().copy().items()

for key, value in snapshot:
    print(f"{key:<10} => {value}")
```

**test_module/__init__.py**:
```python
from test_module.a import value

__all__ = ("value",)
```

**test_module/a.py**:
```python
value = 42
```

Run with python3:
```
esp32@ganymed:~/micropython/ports/unix/build-standard$ python3 ~/run_test_module.py
__name__   => __main__
__doc__    => None
__package__ => None
__loader__ => <_frozen_importlib_external.SourceFileLoader object at 0x7f2f4d25afd0>
__spec__   => None
__annotations__ => {}
__builtins__ => <module 'builtins' (built-in)>
__file__   => /home/esp32/run_test_module.py
__cached__ => None
value      => 42
```

Running with micropython (unix-port):
```
esp32@ganymed:~/micropython/ports/unix/build-standard$ ./micropython ~/run_test_module.py
a          => <module 'test_module.a' from '/home/esp32/test_module/a.py'>
value      => 42
__name__   => __main__
__file__   => /home/esp32/run_test_module.py
```

Running with micropython (esp32S3):
```
MPY: soft reboot
a          => <module 'test_module.a' from 'test_module/a.py'>
value      => 42
__name__   => run_test_module
__file__   => run_test_module.py
MicroPython v1.24.0-preview.199.g0a11832cd.dirty on 2024-08-15; Generic ESP32S3 module with Octal-SPIRAM with ESP32S3
Type "help()" for more information.
```
On the ESP32S3 I imported **run_test_module**, which then executes the code in **run_test_module**.
So it seems to work, but there are differences between **CPython** and **micropython**.

**Comment by shariltumin** [comment]
```bash
$ tree .
.
├── pkga
│   └── pkgb
│       ├── __init__.py
│       ├── M1.py
│       └── pkgc
│           ├── __init__.py
│           └── M2.py
└── test.py

3 directories, 5 files
```

"pkga/pkgb/\_\_init\_\_.py"
```py
M1=__import__("M1", None, None, True, 1)
globals()["add"]=M1.add
globals()["prod"]=M1.prod
```

"pkga/pkgb/M1.py"
```py
def add(a,b):
    return a+b
def prod(a,b):
    return a*b
```

"pkga/pkgb/pkgc/\_\_init\_\_.py"
```py
M2=__import__("M2", None, None, True, 1)
globals()["minus"]=M2.minus
globals()["div"]=M2.div
```

"pkga/pkgb/pkgc/M2.py"
```py
def minus(n,m):
    return n-m

def div(n,m):
    return n/m
```

"test.py"
```py
import sys
print(sys.implementation)

from pkga.pkgb import *
from pkga.pkgb.pkgc import *

print('add(2,3) =', add(2,3))
print('prod(2,3) =', prod(2,3))
print('minus(5,3) =', minus(5,3))
print('div(5,3) =', div(5,3))
```

Run the test:
```bash
$ mpremote mount . exec "import test"
Local directory . is mounted at /remote
(name='micropython', version=(1, 24, 0, 'preview'), _machine='WaveShare Pico Zero with RP2040', _mpy=4870)
add(2,3) = 5
prod(2,3) = 6
minus(5,3) = 2
div(5,3) = 1.666667
```

**Comment by jhugery-blade** [comment]
I think all points are being missed.  I have the following folder structure:

/--
|
|--/lib
     |
     |- modulea.py
     |-moduleb.py
     |-modulec.py

What I want to do is from a script in the root load whatever libraries are in the /lib folder without having to specify the individual library in the directory.
In python3 I would run the following:
   from os.path import dirname, basename, isfile, join
   import glob
   modules = glob.glob(join(dirname(__file__), "*.py"))
   __all__ = [ basename(f)[:-3] for f in modules if isfile(f) and not f.endswith('__init__.py')]

THis is not possible in micropython as glob is not a valid module.  Any ideas?

**Comment by bixb922** [comment]
```os.listdir(folder)``` returns a list with all files in that folder. You can then filter with ```.endswith(".py")``` and import the required modules with ```__import()__```. To import the modules, the folder must be in the list ```sys.path```
